SampleViewer/DownloadsManager.cpp

```diff
@@ -0,0 +1,933 @@
+}
+
+bool DownloadsManager::isSampleDownloading(Sample* sample) const
+{
+  if (!sample || sample->dataItems()->size() == 0)
+    return false;
+
+  const int totalItems = sample->dataItems()->size();
+  for (int j = 0; j < totalItems; ++j)
+  {
+    DataItem* dataItem = sample->dataItems()->at(j);
+    QString key = dataItem->itemId() + "|" + m_homePath + dataItem->path().mid(1);
+    if (m_dataItemProgress.contains(key))
+      return true;
+  }
+  return false;
+}
+
+void DownloadsManager::deleteDataItemFile(DataItem* dataItem)
+{
+  if (!dataItem)
+    return;
+
+  QJsonObject tracking = loadDownloadTracking();
+
+  if (!tracking.contains(dataItem->itemId()))
+  {
+    // qWarning() << "No tracking data found for item:" << dataItem->itemId();
+    return;
+  }
+
+  QJsonObject itemData = tracking[dataItem->itemId()].toObject();
+  QJsonArray filesArray = itemData["files"].toArray();
+
+  QStringList deletedFiles;
+  QSet<QString> deletedDirs;
+
+  for (int i = filesArray.size() - 1; i >= 0; --i)
+  {
+    QString filePath = filesArray[i].toString();
+    QFileInfo fileInfo(filePath);
+
+    if (fileInfo.exists())
+    {
+      if (fileInfo.isDir())
+      {
+        QDir dir(filePath);
+        if (dir.removeRecursively())
+        {
+          deletedDirs.insert(filePath);
+        }
+        else
+        {
+          qWarning() << "Failed to delete directory:" << filePath;
+        }
+      }
+      else if (fileInfo.isFile())
+      {
+        if (QFile::remove(filePath))
+        {
+          deletedFiles.append(filePath);
+        }
+        else
+        {
+          qWarning() << "Failed to delete file:" << filePath;
+        }
+      }
+    }
+  }
+
+  // Clean up empty parent directories
+  QString dataRoot = m_homePath + "/ArcGIS/Runtime/Data";
+  QSet<QString> parentPaths;
+
+  for (const QString& filePath : deletedFiles)
+  {
+    parentPaths.insert(QFileInfo(filePath).dir().absolutePath());
+  }
+
+  for (const QString& dirPath : deletedDirs)
+  {
+    parentPaths.insert(QFileInfo(dirPath).dir().absolutePath());
+  }
+
+  QStringList sortedPaths = parentPaths.values();
+  std::sort(sortedPaths.begin(), sortedPaths.end(), [](const QString& a, const QString& b) {
+    return a.length() > b.length();
+  });
+
+  for (const QString& parentPath : sortedPaths)
+  {
+    cleanupEmptyParentDirectories(parentPath, dataRoot);
+  }
+
+  // Remove tracking data
+  tracking.remove(dataItem->itemId());
+  saveDownloadTracking(tracking);
+}
+
+void DownloadsManager::cancelDataItem(const QString& dataItemKey)
+{
+  PortalItem* portalItem = nullptr;
+  if (m_activeDownloads.contains(dataItemKey))
+    portalItem = m_activeDownloads[dataItemKey];
+
+  m_dataItemProgress.remove(dataItemKey);
+  m_activeDownloads.remove(dataItemKey);
+
+  if (portalItem)
+    portalItem->cancelLoad();
+}
+
+void DownloadsManager::cleanupEmptyParentDirectories(const QString& startPath, const QString& boundary)
+{
+  QDir currentDir(startPath);
```

Comment:
optional, but consider braced initializer syntax when possible.

```suggestion
  QDir currentDir{startPath};
```